### ISSUE TITLE
LG-284: MicroProfile OpenTracing 2.0 Implementation (part of MP 4.0)

### ISSUE THREAD

**Issue by donbourne** [OpenLiberty/open-liberty#11474]
Describe the high level feature, including any external spec links.

When ready, add links to the Upcoming Feature Overview document and Feature Test Summary issue:
- UFO: https://ibm.box.com/s/yc1j94zht3kembig2hpph9z76jg96dc0
- Guide: https://github.com/OpenLiberty/guide-microprofile-opentracing/issues/126
- ID: https://github.com/OpenLiberty/docs/issues/3290
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [x] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [x] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [x] Complete any follow-ons from the POC Review.
- [x] Design / UFO Approval (Alasdair Nottingham) or N/A.
- [ ] No Design / No UFO Approval (Arthur De Magalhaes - cloud / Alasdair Nottingham - server) or N/A.
- [x] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [x] ID Requirements identified. (Epic owner / Feature owner with ID focal point)
- [x] Create a child task of this epic entitled "FAT Approval Test Summary". Add the link in above.

## **Legal**
3 weeks before Onboarding
- [x] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.

## **Translation**
3 weeks before Onboarding
- [x] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [x] Implementation complete. (Epic owner / Feature owner)
- [x] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [ ] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [x] **FAT** - (Kevin Smith). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [x] **Demo** - (Tom Evans or Chuck Bridgham). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [x] **Globalization** (Sam Wong - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [x] **Accessibility** - (Steven Zvonek). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [x] **ID** - (Kareen Deen). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [x] **Performance** - (Jared Anderson). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [x] **Serviceability** - (Don Bourne). Serviceability has been addressed.
- [x] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [x] **SVT** - (Brian Hanczaryk). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.

## **Ready for GA**
1 week before Onboarding
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)

## **Other deliverbles**
- [x] **OL Guides** - (Yee-Kang Chang). Assessment for OL Guides is complete or N/A.
- [x] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
- [ ] **Blog** - (Laura Cowen) Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by donbourne** [comment]
UFO review comments

```
p. 4
    ACTION - toTraceId / toSpanId not breaking changes... add another heading
    
p. 14
    how can you make zipkin work if it only supports 0.32?
        0.33 just removes deprecated functions from 0.32, otherwise compatible.  zipkin works with 0.33
        ACTION - should indicate that toTraceId / toSpanId came in 0.32 in the UFO
    
    ACTION - discuss offline    
        can we support zipkin in same way as jaeger (instead of user feature?)
            would need a separate feature to do that for zipkin
            if zipkin is still important should be supported as a library rather than a user feature
            may not be worth doing if opentelemetry supports export to zipkin and jaegar
        
        opentelemetry
            built in exporter to zipkin and jaeger
        
        alasdair - not convinced we should update zipkin tracer 
            felix - that's relatively easy

    tom evans - how can we pick up service fixes if OT keeps changing even in minor releases?        
        
john alcorn
    dev experience - depend on jaeger client
    will we update docs to talk about easier dev experience approach (maven dependency approach rather than long list of jars in shared lib)?
        ACTION - see if docs are up to date on that
    
emily - do we have to specify 3rd party still on the api?
    only when you want to use the opentracing api directly - if not, you can just run without having 3rd party
    that's not what emily read in release notes
    ACTION - docs say to add 3rd party to avoid problems - reword to show that this is optional    

p. 27
    ACTION - alasdair - would like perf test on this. just say on this page that perf team will do normal regression test
    
p. 30
    ACTION - need to discuss offline merits of doing 9874 in this feature
    
p. 32
    emily - is there any migration doc for going to OT 0.33?
        covered in a medium post and in the javadoc
        
p.33
    how much progress on openTelemetry in smallrye?
    ACTION - alasdair - this page isn't relevant to the design - move to backup and change it so it doesn't say it's a migration impact or remove it
    
emily
    ACTION - when updating the point about optional 3rd party setting, also update the part about using maven dependency
```

**Comment by fmhwong** [comment]
Updated UFO and uploaded to box folder

**Comment by fmhwong** [comment]
@steven1046 Hi Steve, this feature does not have UI.  May I have your accessibility approval?

**Comment by fmhwong** [comment]
@samwatibm Hi Sam, this feature reuses the messages from previous version. No new messages are introduced.  May I have your globalization approval?

**Comment by fmhwong** [comment]
@skasund STE for this feature has been uploaded to box folder. Please approve for STE. Thanks!

**Comment by chirp1** [comment]
I indicated to Felix that the capabilities for this feature would be part of a larger discussion on distributed tracing that will be documented in the new year. Felix opened this issue for ID and added the information to document for this particular feature. https://github.com/OpenLiberty/docs/issues/3290 Felix will do a blog as usual so that customers have enough information for the customer to use the function until we get the concept/reference documentation written.

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

5.  Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?

**Comment by ayoho** [comment]
FYI: I've removed the FAT approval for now since the feature is not yet using the final version of the MicroProfile API. See:
- https://github.com/OpenLiberty/open-liberty/blob/integration/dev/cnf/oss_dependencies.maven#L381
- https://github.com/OpenLiberty/open-liberty/blob/integration/dev/io.openliberty.org.eclipse.microprofile/opentracing.2.0.bnd#L23

Once the final API is made available and the feature has been updated to use it, I can re-add the approval (also assuming all functionality remains in good working order).

**Comment by fmhwong** [comment]
@ayoho FYI: PR https://github.com/OpenLiberty/open-liberty/pull/15396 has been merged.

**Comment by donbourne** [comment]
@fmhwong , please add comment with answers for serviceability approval (even if we don't have SVT complete yet it's helpful for seeing if everything else is ready for this approval)

**Comment by fmhwong** [comment]
Serviceability Approval Comment - Please answer the following questions for serviceability approval:

1. WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR? Have these issues been addressed in the implementation? **Yes.  The main change was the OpenTracing API upgrade to 0.33.0 so no usage scenarios have changed.**

1. Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).
    1. What problem paths were tested and demonstrated?  **No new problem path**
    1. Who did you demo to?  **At end of iteration call on 12/15/2020**
    1. Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?  **No questions or concerns from the attendees**

1. SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them. Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
    1. Who conducted SVT tests for this feature?  **Dan Guinan** 
    1. Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  **No concerns so far**

1. Which L2 / L3 queues will handle PMRs for this feature? Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it. Ask Don Bourne if you need links or more info.  **WAS L3: Liberty Log Analytics**

1. Does this feature add any new metrics or emit any new JSON events? If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?  **No new metrics or JSON events**

**Comment by donbourne** [comment]
Thanks @fmhwong , given the lack of changes I don't have any concerns.  Just waiting for SVT's comment on 3b.
